### ISSUE TITLE
Bug: Incomplete optional chaining resulting in prop access on undefined error in FunctionScope.ts

### ISSUE THREAD

**Issue by mjavaid** [typescript-eslint/typescript-eslint#7562]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

scope-manager

### Repro Code

```typescript
// input.ts file
class AuthService {
  getUserByToken(_: any) { throw new Error("Function not implemented."); }
}

function appInitializer(authService: AuthService) {
  return () => {
    return new Promise((_) => {
      authService.getUserByToken(_);
    });
  };
}


// parser.js file
const parser = require('@typescript-eslint/parser');
parser.parse('input.ts', { range: false });
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Expected ESTree to be generated without issue.

### Actual Result

Instead the following error generated:

```
@typescript-eslint\scope-manager\dist\scope\FunctionScope.js:28
        const bodyStart = this.block.body?.range[0] ?? -1;
                                                ^

TypeError: Cannot read properties of undefined (reading '0')
```

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `NA` |
| `@typescript-eslint/parser`            | `6.5.0` |
| `@typescript-eslint/rule-tester`       | `NA` |
| `@typescript-eslint/scope-manager`     | `6.5.0` |
| `@typescript-eslint/typescript-estree` | `6.5.0` |
| `@typescript-eslint/type-utils`        | `NA` |
| `@typescript-eslint/utils`             | `NA` |
| `TypeScript`                           | `5.2.2` |
| `ESLint`                               | `8.48.0` |
| `node`                                 | `16.18.0` |

**Comment by bradzacher** [comment]
`range: false`

This is expected to fail if you pass this option.

 It's crashing because the data it expects is not there because you turned it off.

**Comment by mjavaid** [comment]
> `range: false`
> 
> This is expected to fail if you pass this option.
> 
> It's crashing because the data it expects is not there because you turned it off.

Hi @bradzacher ,thanks for the response! Two points to note:

1. The default value for `range` in this scenario was `false`. I had defined it explicitly in the repro example so that any other configs are overwritten.
2. There is a nullish coalescing operator (`??`) in `FunctionScope.ts` that appears to be there to gracefully handle a missing `range` prop, but this expression fails as the optional chaining does not extend to the `range[0]`
```
this.block.body?.range[0] ?? -1;
```

If the expected behavior is a _fail_ condition then the null check seems redundant?

**Comment by bradzacher** [comment]
> The default value for range in this scenario was false

Correct. Though it's worth noting that our parser package is not intended to be called programmatically - it's a parser for ESLint intended to be called how ESLint calls a parser.

I.e. You need to pass the right arguments to it and match ESLint.

This is why [the package docs](https://typescript-eslint.io/packages/parser) do not contain any examples of how to call it programmatically.

If you want to use our parser programmatically, use our [`typescript-estree` package](https://typescript-eslint.io/packages/typescript-estree) instead.

---

> There is a nullish coalescing operator (??) in FunctionScope.ts that appears to be there to gracefully handle a missing range prop

You've misunderstood the code - that operator is there to handle the case where `this.block.body` is nullish.

The code assumes the range always exists as that is the contract that ESLint defines.

**Comment by mjavaid** [comment]
> > The default value for range in this scenario was false
> 
> Correct. Though it's worth noting that our parser package is not intended to be called programmatically - it's a parser for ESLint intended to be called how ESLint calls a parser.
> 
> I.e. You need to pass the right arguments to it and match ESLint.
> 
> This is why [the package docs](https://typescript-eslint.io/packages/parser) do not contain any examples of how to call it programmatically.
> 
> If you want to use our parser programmatically, use our [`typescript-estree` package](https://typescript-eslint.io/packages/typescript-estree) instead.
> 
> ---
> 
> > There is a nullish coalescing operator (??) in FunctionScope.ts that appears to be there to gracefully handle a missing range prop
> 
> You've misunderstood the code - that operator is there to handle the case where `this.block.body` is nullish.
> 
> The code assumes the range always exists as that is the contract that ESLint defines.

Thanks for the clarification here and for answering my questions! I will look into switching to the `typescript-estree` package instead for my use case 👍

I hadn't been able to find many examples for a programmatic way of achieving what I wanted to do so the pointer helps a lot.
